### DISCUSSION TITLE
Native install instructions for MSP430 gcc

### DISCUSSION THREAD

**Discussion by jbrusey** [contiki-ng/contiki-ng#2189] [Show and tell]
https://github.com/contiki-ng/contiki-ng/wiki/Toolchain-installation-on-Linux#install-msp430-compiler gives instructions for installing a 32-bit binary or a script for compiling on older versions of Linux. 

I have just been through the process of creating a revision of that script that seems to work for newer versions of Linux. 

https://gist.github.com/jbrusey/441951b5ed3565a6d1c15dfc3b64f601

What's the process to get this through to the documentation as I think it is probably more useful than the version that is there currently?

**Comment by pjonsson** [comment]
The wiki has been moved into the source tree so the file linked is `doc/getting-started/Toolchain-installation-on-Linux.md`.

Make a pull request with the updated documentation and the script from the gist. I suggest putting the script under `tools/toolchains/msp430`.

The comment on line 497 in the script refers to gcc 4.7.0, but the script is building 4.7.2. The error checking can be simplified by adding `-e` as argument to `sh` on the first line, and this kind of helper:

```bash
error() {
    echo "$*"
    exit 1
}

[ -d directory ] || error "directory does not exist"
```

And the issue with a used temporary directory can be avoided with using `mktemp` to create a fresh directory:
```bash
tmpdir=$(mktemp -d -p "$PWD")
```

**Reply by jbrusey** [reply]
Thanks for this feedback. 

The line on 497 that refers to 4.7.0 is due to the patch seemingly only being available for 4.7.0 and not 4.7.2. I notice that it fails on a number of counts but, prior to using sh -e, I didn't notice it. I can ignore that error for the moment but perhaps it is worth fixing? Looking at the rejects from the patch, I don't think the answer will be easy. 

Is there any plans to move to using the more recent version of msp430-elf-gcc?

**Reply by pjonsson** [reply]
I guess the patch issues aren't a recent thing, just add a comment that some failures are expected. I had a second look at the gist and there's also a "compiping" comment on line 503.

MSP430 is the only Contiki-NG target with an "old" compiler, so it would be nice to have a more modern compiler. The GCC 9 based MSP430-compiler from Texas Instruments seems to work for Contiki-NG, but the newlib-based libc that comes with it takes up too much memory. Recompiling that libc with additional slimline options makes it possible to get hello world working, but the other MSP430 regression tests still become too big. Using the current libc isn't a quick fix either because it contains some assembly and the calling convention was changed.

I think the way forward for a GCC update would be to port picolibc to MSP430. Picolibc is based on newlib, so with a bit of luck it might not be that much work to bring back the MSP430 support to it, even if it is described as "designed for embedded 32- and 64-bit systems".

**Reply by jbrusey** [reply]
Just to complete / close this thread - the pull request is https://github.com/contiki-ng/contiki-ng/pull/2191
